contracts/token/ERC20/ERC20Permit.sol

```diff
@@ -0,0 +1,109 @@
+pragma solidity ^0.6.0;
+
+import "./ERC20.sol";
+import "./IERC2612Permit.sol";
+import "../../cryptography/ECDSA.sol";
+import "../../utils/Counters.sol";
+
+/**
+ * @dev Extension of {ERC20} that allows token holders to use their tokens
+ * without sending any transactions by setting {IERC20-allowance} with a
+ * signature using the {permit} method, and then spend them via
+ * {IERC20-transferFrom}.
+ *
+ * The {permit} signature mechanism conforms to the {IERC2612Permit} interface.
+ */
+abstract contract ERC20Permit is ERC20, IERC2612Permit {
+    using Counters for Counters.Counter;
+
+    mapping (address => Counters.Counter) private _nonces;
+
+    bytes32 private immutable _PERMIT_TYPEHASH = keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+
+    // Mapping of ChainID to domain separators. This is a very gas efficient way
+    // to not recalculate the domain separator on every call, while still
+    // automatically detecting ChainID changes.
+    mapping (uint256 => bytes32) private _domainSeparators;
```

Comment:
What’s the gas benefit? I don’t see it. I have done some tests on Remix and it seemed to me that it was slightly more expensive. Maybe I’m missing something.
This is what I tried on Remix, and the execution cost of the first one was 1156 while the second one was 810. Besides the SSTORE used in the constructor.

```
pragma solidity 0.6.11;


contract DomainSeparator1 {
    string constant internal NAME = "LUSD Stablecoin";
    string constant internal VERSION = "1";

    mapping (uint256 => bytes32) private _domainSeparators;

    constructor() public {
        _updateDomainSeparator();        
    }

    function domainSeparator() external returns (bytes32) {
        bytes32 domainSeparator_ = _domainSeparators[_chainID()];
        if (domainSeparator_ != 0x00) {
            return domainSeparator_;
        } else {
            return _updateDomainSeparator();
        }
    }

    function _updateDomainSeparator() private returns (bytes32 domainSeparator_) {
        uint256 chainID = _chainID();
        domainSeparator_ = keccak256(abi.encode( 
            keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'),
            keccak256(bytes(NAME)), keccak256(bytes(VERSION)), chainID, address(this)));
            
        _domainSeparators[chainID] = domainSeparator_;
    }

    function _chainID() private pure returns (uint256 chainID) {
        assembly {
            chainID := chainid()
        }
    }
}


contract DomainSeparator2 {
    string constant internal NAME = "LUSD Stablecoin";
    string constant internal VERSION = "1";

    mapping (uint256 => bytes32) private _domainSeparators;
    
    function domainSeparator() external view returns (bytes32) {
        uint256 chainID = _chainID();
        return keccak256(abi.encode( 
            keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'),
            keccak256(bytes(NAME)), keccak256(bytes(VERSION)), chainID, address(this)));
        
    }

    function _chainID() private pure returns (uint256 chainID) {
        assembly {
            chainID := chainid()
        }
    }
}
```

Reply:
@bingen Thanks for bringing this up. It does seem that the version that doesn't use storage is cheaper.

@nventuro Has something change in the costs of storage that this is now cheaper?

Reply:
I think you can remove `_domainSeparators ` in `DomainSeparator2`

Reply:
I've done some tests and concluded that the storage based approach in this PR is indeed inferior, and I've arrived at a better solution using a combination of `immutable` and `keccak` as fallback for when chain id has changed.

https://github.com/frangio/domain-separator-gas-report/blob/311306300546de7a021b856fb2ff1723d5833a34/contracts/Test.sol#L86-L92

Reply:
> I think you can remove _domainSeparators in DomainSeparator2

Yes, indeed, thanks, it was just a quick test in Remix, but I don’t think it’s relevant for the gas comparison.